stellargraph/layer/graph_attention.py

```diff
@@ -0,0 +1,481 @@
+                head
+            ]  # Attention kernel a in the paper (2F' x 1)
+
+            # Compute inputs to attention network
+            features = K.dot(X, kernel)  # (N x F')
+
+            # Compute feature combinations
+            # Note: [[a_1], [a_2]]^T [[Wh_i], [Wh_2]] = [a_1]^T [Wh_i] + [a_2]^T [Wh_j]
+            attn_for_self = K.dot(
+                features, attention_kernel[0]
+            )  # (N x 1), [a_1]^T [Wh_i]
+            attn_for_neighs = K.dot(
+                features, attention_kernel[1]
+            )  # (N x 1), [a_2]^T [Wh_j]
+
+            # Attention head a(Wh_i, Wh_j) = a^T [[Wh_i], [Wh_j]]
+            dense = attn_for_self + K.transpose(
+                attn_for_neighs
+            )  # (N x N) via broadcasting
+
+            # Add nonlinearity
+            dense = LeakyReLU(alpha=0.2)(dense)
+
+            # Mask values before activation (Vaswani et al., 2017)
+            # YT: this only works for 'binary' A, not for 'weighted' A!
+            # YT: if A does not have self-loops, the node itself will be masked, so A should have self-loops
+            # YT: this is ensured by setting the diagonal elements of A tensor to 1 above
+            mask = -10e9 * (1.0 - A)
+            dense += mask
+
+            # Apply softmax to get attention coefficients
+            dense = K.softmax(dense)  # (N x N), Eq. 3 of the paper
+
+            # Apply dropout to features and attention coefficients
+            dropout_feat = Dropout(self.in_dropout_rate)(features)  # (N x F')
+            dropout_attn = Dropout(self.attn_dropout_rate)(dense)  # (N x N)
+
+            # Linear combination with neighbors' features [YT: see Eq. 4]
+            node_features = K.dot(dropout_attn, dropout_feat)  # (N x F')
+
+            if self.use_bias:
+                node_features = K.bias_add(node_features, self.biases[head])
+
+            # Add output of attention head to final output
+            outputs.append(node_features)
+
+        # Aggregate the heads' output according to the reduction method
+        if self.attn_heads_reduction == "concat":
+            output = K.concatenate(outputs)  # (N x KF')
+        else:
+            output = K.mean(K.stack(outputs), axis=0)  # N x F')
+
+        output = self.activation(output)
+        return output
+
+    def compute_output_shape(self, input_shape):
+        output_shape = input_shape[0][0], self.output_dim
+        return output_shape
+
+
+class GAT:
+    """
+    A stack of GAT layers with aggregation of multiple attention heads, Eqs 5-6 of GAT paper
+    """
+
+    def __init__(
+        self,
+        layer_sizes,
+        attn_heads=1,
+        attn_heads_reduction=None,
+        activations=None,
+        bias=True,
+        in_dropout=0.0,
+        attn_dropout=0.0,
+        normalize="l2",
+        generator=None,
+    ):
+        """
+
+        Args:
+            layer_sizes: list of output sizes of GAT layers in the stack
```

Comment:
It's a good idea to specify the accepted types of the arguments. For example (list of int) or (List[int]).

Reply:
fixed